proposals/1141_audit-logging/README.md

```diff
@@ -0,0 +1,323 @@
+logs.
+
+### How Pinniped Works Today (as of version v0.16.0)
+
+The Pinniped Supervisor and Concierge components are Kubernetes Deployments. Today, each Pod has a single container,
+which is the Supervisor or Concierge app. Kubernetes captures the stdout of the app into the Pod logs.
+
+Today, the Pinniped Supervisor and Concierge log many interesting events to their Pod logs. These logs are meant
+primarily to help an admin user debug problems with their Pinniped configuration or with their cluster. The Supervisor
+and Concierge each offer an install-time configuration option to turn up the verbosity of these Pod logs.
+
+However, these logs are not meant to be audit logs. They generally focus on logging problems, not on logging successes.
+They try to avoid logging anything that might be confidential or PII (personally identifiable information). Since email
+addresses might be considered PII, these logs generally avoid including usernames at the default log level, since
+usernames could be email addresses in some configurations. Logging the identity of actors (usernames) are a key aspect
+of audit logs.
+
+## Terminology / Concepts
+
+None.
+
+## Proposal
+
+The goal of an audit log is to log events that could be helpful in a forensic investigation of past usage, including the
+actor (the username) and the actions that were taken on the system.
+
+### Goals and Non-goals
+
+Goals
+
+- Auditing events relating to upstream identity provider (IDP) authentication, refresh, and sessions.
+- Auditing events relating to minting and validating cluster credentials.
+- Enabling auditors to easily stitch together authentication events into an audit trail.
+- Provide consistent data across auditable events.
+- Provide the ability to enable and disable auditing.
+- Provide the ability to route audit logs to a separate destination from the rest of Pinniped’s logs.
+
+Non-goals
+
+- Enabling auditing in the impersonation proxy. If needed, this will be handled in a separate feature.
+- Providing the ability to filter or choose which audit events to capture.
+- Auditing the management of CRs (e.g. OIDCIdentityProvider). These events are captured by the API server audit logs.
+
+### Specification / How it Solves the Use Cases
+
+This proposal recommends following the recommendation of the Kubernetes docs to create a separate Pod container log.
+This new container log will contain the audit logs (and only the audit logs).
+
+#### API Changes
+
+##### Configuration Options
+
+There will be very few user-facing configuration options for audit logging in the first version of the feature. If later
+found to be needed, more configuration could be added in future versions.
+
+This proposal recommends adding a single on/off install-time configuration option for disabling audit logs. By default,
+audit logs will be enabled. An admin user who is concerned about logging identities, for example because usernames may
+be considered PII, may disable audit logging.
```

Comment:
That's a fair point, although it might be a questionable practice for admins to expose these logs to non-admin users in the first place.

My concern in disabling by default is that users who install using our pre-rendered YAML that we ship with every release would then have audit logs turned off, making it possible that many installations would not benefit at all from this new feature.

If we feel that audit logging is a best security practice, then I'd suggest that we enable it by default. Clearly we could go either way though, since we intend for it to be configurable on/off.

Reply:
> My concern in disabling by default is that users who install using our pre-rendered YAML that we ship with every release would then have audit logs turned off, making it possible that many installations would not benefit at all from this new feature.

Audit logging is meaningless if no one is looking, i.e. you need tooling to actually consume and alert on events.  So turning it on by default does not provide value on its own.  I do not expect our pre-rendered YAML to realistically support all production environments, but I do expect them to always be safe.  Since the proposal attempts to re-use pod logs as the audit mechanism, it has the potential to leak PII into a stream that may not expect that to happen.  Thus we need users to opt-in.

Reply:
> Audit logging is meaningless if no one is looking, i.e. you need tooling to actually consume and alert on events.

A primary benefit of audit logging is to have the logs off-boarded in near-realtime and archived somewhere tamper-proof, such that they can be referenced during the investigation of an ongoing or past security incident. Any cluster which is already offloading and storing pod logs using something like fluentbit would automatically gain this benefit upon upgrade if the Pinniped audit logs were enabled by default.

The benefits of enabling by default should be weighed against the potential downside that you mentioned.

Reply:
Updated in 3cf3b28c5b0a24b5f9adf08fb40cb0716e42034e.